[MM-25435] Emoji picker + skin tones

Willyfrog opened the pull request:
#### Summary

Redesigned emoji picker + skin tone selector

it is not finished, but functionality is there, I have yet to go over the design files again and see what I've missed. but I'd like a first review to see if there is anything that needs to change

##### Known issues:
- it takes a bit to load the first time, faster afterwards
- some emojis seems to not be properly clasified as "skinnable", I have to go over those and see if they have any differences from the others (e.g: family emojis)

#### Ticket Link

[MM-25435](https://mattermost.atlassian.net/browse/MM-25435)

#### Related Pull Requests

#### Screenshots

<img width="333" alt="Screen Shot 2021-06-18 at 15 35 56" src="https://user-images.githubusercontent.com/1515906/122569298-e4388a00-d04a-11eb-9fc2-5c6bd11d782b.png">

#### Release Note

```release-note
Redesigned Emoji Picker & skin selector
```

esethna commented:
Thanks @Willyfrog! Looking good, a few things I noticed:

1. If you select a few different default skin tones then execute a search it seems to return a bunch of different options for each skin tone:
![image](https://user-images.githubusercontent.com/14333569/122593852-15996180-d01b-11eb-9933-01af9862ae7e.png)

2. Didn't test this but just confirming, the "Custom Emoji" button will only appear if the user has permissions to access the custom emoji page, correct?

@matthewbirtch can you help with a priority review on this?

esethna reviewed (approved):
Approving as none of the issues I listed are blockers

matthewbirtch reviewed (changes requested):
Thanks @Willyfrog. This is a good start. I did a first pass review and there are definitely a number of tweaks to bring this up to the spec with the new emoji picker design. I'll try to outline it as clearly as I can. There may be a few more things, but this is all I had time for today, so I thought I would get this out to you.

### Emoji Picker Positioning
**From Message Input**
- Align the right edge of the popover with the right edge of the button that triggers it
- The bottom of the popover should be 4px up from the button that triggers it
![image](https://user-images.githubusercontent.com/2040554/122596646-2e644000-d038-11eb-87d5-4cb77388201b.png)

**From the Message Context**
- Align the *left* edge of the emoji picker with the left edge of the button that triggers it
- If the emoji picker displays above the reaction button, it should be 4px above the top of the button that triggers it
- If the emoji picker displays below the reaction button, it should be 4px below the bottom of the button that triggers it
![image](https://user-images.githubusercontent.com/2040554/122597039-ceba6480-d038-11eb-9bf5-710f0c5df4d0.png)

**From the Message Hover Actions Context**
- Align the right edge of the emoji picker with the right edge of the button that triggers it
- If the emoji picker displays above the reaction button, it should be 4px above the top of the button that triggers it
- If the emoji picker displays below the reaction button, it should be 4px below the bottom of the button that triggers it

### UI tweaks

**Main Container**
- The popover container should have a `box-shadow: 0px 8px 24px rgba(0, 0, 0, 0.12);` on it (also known as "Elevation 4" in our design system)
- The Container's radius should be `border-radius: 4px;`

**Tabs**
- The `font-size` for 'Emojis' should be 12px and the `font-weight` should be 600
- There needs to be a divider line directly under the Emojis/gif Tabs
- The first tab should be 12px from the left edge. It looks to be 5px right now

**Category Icons**
- The category icon buttons should have `width: 28px;` and `height: 28px;`
- The icons within the buttons should have `font-size: 18px;`
- The active category icon should be `color: var(--button-bg);`. It looks like it's getting the channel color instead for some reason
- For the hover state of the category icon buttons, the icons should be `color: var(--center-channel-color-72);`
}

**Search and Skin Tone Area**
- Can we use the compass `icon-magnify` instead of the font awesome icon for the search input?
- The search input should only have the blue outlined in the focused state
- The skin tone button should have a hover state. The [design file here](https://www.figma.com/file/GPTxUI7ACBbbdRfM8Dbtl6/MM-25435-Emoji-Picker-UI-Refresh?node-id=13%3A2667) shows the hover state. Sample below as well
![image](https://user-images.githubusercontent.com/2040554/122590227-408db080-d02f-11eb-8ab3-696808664fa9.png)
- The skin tone button's tooltip is displaying behind the popover
- The words `Default skin tone` should have `line-height: 14px;`
- When all the skin tones are showing, each of the skin tone buttons should have a hover state as demonstrated in the [design here](https://www.figma.com/file/GPTxUI7ACBbbdRfM8Dbtl6/MM-25435-Emoji-Picker-UI-Refresh?node-id=13%3A3794)
- It seems like it doesn't always choose the skin tone that I've clicked (see gif below).
![2021-06-18 13 16 48](https://user-images.githubusercontent.com/2040554/122596232-8d758500-d037-11eb-96fe-e08e0563735c.gif)
- Is it possible to animate the skin tones in (slide from right to left) as they are shown and out (slide from right to left) as they are hidden?

**Emojis List**
- The emoji picker list should have `padding: 4px 0 4px 12px;`
- With the above, you'll need to adjust the category headers and remove the -6px left position
- Each emoji in the list should have container that is `width:36px;` and `height: 36px;` and no margin between items
- Design had 9 emojis in each row

**Footer**
- The text label is misaligned vertically in the footer
![image](https://user-images.githubusercontent.com/2040554/122597510-73d53d00-d039-11eb-8607-39e89d848d5f.png)
  - Everything should line up like this:
![image](https://user-images.githubusercontent.com/2040554/122595924-22c44980-d037-11eb-86f9-3266debbe880.png)
- The `Custom Emoji` button should have `font-weight: 600` and `font-size: 12px;`

**Mobile Webview**
- This will need some work, but I didn't have time to go through it in detail yet. Let's address what we can above first.

Willyfrog commented:
@matthewbirtch thanks for the detailed review.

for the positioning, if it goes within the text input box usually is because the user resized the window after displaying the emoji picker or the window doesn't fit the emoji picker. I haven't done anything to the positioning of the emoji picker, but I agree it should improve. Would it be ok to add a ticket to take a look at it?

I've started working on the rest of the items, the only one that worries me a bit is the missalignment of the text name of the emoji. when there is more than 1 alias it displays in two lines, if I center the label, the second one gets cut. I'll try to take another look in the morning with fresher eyes.

matthewbirtch commented:
Thanks @Willyfrog. I'm totally ok with a follow up ticket for the positioning. We can leave that out of this PR if you'd prefer

jwilander commented on actions/emoji_actions.js:
```diff
@@ -43,6 +43,16 @@ export function incrementEmojiPickerPage() {
 }
 
+export function setRecentSkin(skin) {
+    return async (dispatch) => {
```
Does this need to be async?

Willyfrog replied:
no need, removing `async`

jwilander reviewed (approved):
LGTM - just one non-blocking comment

jwilander commented:
@jgilliam17 are you able to prioritize this today?

Willyfrog commented:
> It seems like it doesn't always choose the skin tone that I've clicked (see gif below).

I can't repro, but I made some changes so I might have fixed it without noticing

@matthewbirtch  I think I addressed all your points minus the positioning and the animation, I'll create a couple of tickets for those.

> Didn't test this but just confirming, the "Custom Emoji" button will only appear if the user has permissions to access the custom emoji page, correct?

@esethna  It checks if the custom emoji is enabled in the server, I didn't look for specific permissions for users. I'll fix that one now.

>  The autocomplete should use the default skin tone selected in the picker rather than showing all the options:

I'll add a ticket for this.

matthewbirtch commented:
> @matthewbirtch I think I addressed all your points minus the positioning and the animation, I'll create a couple of tickets for those.

Thanks @Willyfrog. Looks like the majority are resolved. I'm still seeing the following though. If you'd prefer to address these in a follow-up ticket as well, i'd be ok with that since it's so close.

- Can we shift the tabs to the left a little but to align with the search input below it?
<img width="298" alt="Screen Shot 2021-06-21 at 9 54 32 AM" src="https://user-images.githubusercontent.com/2040554/122773975-cd767b00-d276-11eb-93fb-ac5d3d51425d.png">

- There appears to be too much padding on the left and right side of the tabs. Can you make sure the left and right padding are 12px?

- The active category icon should be `color: var(--button-bg);`. It looks like it's getting the channel color instead. 
![image](https://user-images.githubusercontent.com/2040554/122773749-9902bf00-d276-11eb-805d-e9693d93817c.png)

- When I click on a category, it seems to be focusing on the search input for some reason?

- Can we add 4px of margin between the magnify icon and the Placeholder copy on the search input?

- The text and the icon within the search input seems to shift when clicking in and out of the input field. It shouldn't jump around

- The Skin Tone button is getting cut off. Also, the hover state for this button should be rounded square, but it's a rectangle shape right now. 
![image](https://user-images.githubusercontent.com/2040554/122774226-0282cd80-d277-11eb-9a61-17caa67927f1.png)

- In the footer, the Emoji is aligned too low and not vertically centered with the rest of the content still.
<img width="380" alt="Screen Shot 2021-06-21 at 9 57 43 AM" src="https://user-images.githubusercontent.com/2040554/122774527-407ff180-d277-11eb-99d3-a0778a461338.png">

- The gif picker layout will need to be updated as well. The design proposed a few updates. Can you create a follow-up ticket for that as well?

Willyfrog commented:
> When I click on a category, it seems to be focusing on the search input for some reason?

removed focus, but it's been working like that for a while now.

>  The gif picker layout will need to be updated as well. The design proposed a few updates. Can you create a follow-up ticket for that as well?

sure!

Willyfrog commented:
Added tickets:

[emoji search + skin tones](https://mattermost.atlassian.net/browse/MM-36598)
[emoji picker positioning](https://mattermost.atlassian.net/browse/MM-36599)
[skin tone animation](https://mattermost.atlassian.net/browse/MM-36600)

matthewbirtch reviewed (approved):
Thanks for making all those changes @Willyfrog! It's looking great.

matthewbirtch commented:
@Willyfrog  We should create another ticket for mobile webview.

![image](https://user-images.githubusercontent.com/2040554/122800378-db38fa00-d290-11eb-956d-80e348301629.png)

esethna commented:
@hmhealey @jgilliam17 can you guys help prioritize a review on this PR and help with merging today?

esethna commented:
@Willyfrog on the latest test server it looks like you have to select the skin tone twice before it takes effect in the picker

jgilliam17 commented:
( forgot to post this before my meetings)
Thanks @Willyfrog 
Beside what Matt pointed out and now Eric, I noticed that skin tone selector needs two clicks to accept the change when opening from "Add a reaction" +, not the main post emoji picker.

Willyfrog commented:
@jgilliam17 @esethna I found the problem, it was storing wrong the key in the state so it took extra long to retrieve it from the localstorage (which is meant to persist across sessions, not for updating the component)

hmhealey commented on components/emoji_picker/emoji_picker.jsx:
```diff
@@ -481,7 +489,12 @@ export default class EmojiPicker extends React.PureComponent {
-    getCategoryByIndex(index) {
+    onSkinSelected = (skin) => {
+        this.props.actions.setRecentSkin(skin);
+        this.props.actions.persistRecentSkin(skin);
```
Also, moving the user's current skin to preferences means we don't need to explicitly persist it

Willyfrog commented on actions/emoji_actions.js:
```diff
@@ -43,6 +43,16 @@ export function incrementEmojiPickerPage() {
     };
 }
 
+export function setRecentSkin(skin) {
```
it is being saved in the local storage, along with the recent emojis, so it persists between sessions but not across devices

Willyfrog commented:
/update-branch

Willyfrog commented:
> @Willyfrog We should create another ticket for mobile webview.
> 
> ![image](https://user-images.githubusercontent.com/2040554/122800378-db38fa00-d290-11eb-956d-80e348301629.png)

https://mattermost.atlassian.net/browse/MM-36613 feel free to add any details, I've set it to triage so we can probably discuss it next sprint planning

jgilliam17 reviewed (approved):
Thanks @Willyfrog 
Tested, looks good to merge.
- Verified re-designed emoji picker and skin tone selector
- The issue with 2 clicks to change the skin tone has been resolved. 
@hmhealey Can session issue you raised be addressed in a follow up PR?

jwilander commented:
/update-branch

hmhealey reviewed (approved):
We're good to switch to Preferences in a separate PR

mattermod commented:
Cherry pick is scheduled.

mattermod commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
Failed to add the RSA host key for IP address '140.82.113.3' to the list of known hosts (/app/.ssh/known_hosts).
From github.com:mattermost/mattermost-webapp
   f8356d11d..962668f1f  cloud      -> upstream/cloud
   c25b41e07..e0e2e2587  master     -> upstream/master
Fetching origin
Failed to add the RSA host key for IP address '140.82.113.3' to the list of known hosts (/app/.ssh/known_hosts).
+++ Updating remotes done...
!!! 'upstream/release-5.37' not found. The second argument should be something like upstream/release-0.21.
    (In particular, it needs to be a valid, existing remote branch that I can 'git checkout'.)

+++ Returning you to the master branch and cleaning up.

```

hmhealey commented:
/cherry-pick cloud

mattermod commented:
Cherry pick is scheduled.
